src/AWS.Deploy.Orchestration/OptionSettingHandler.cs

```diff
@@ -333,5 +349,79 @@ public bool IsSummaryDisplayable(Recommendation recommendation, OptionSettingIte
 
             return true;
         }
+
+        /// <summary>
+        /// Checks whether the option setting item has been modified by the user. If it has been modified, then it will hold a non-default value
+        /// </summary>
+        /// <returns>true if the option setting item has been modified or false otherwise</returns>
+        public bool IsOptionSettingModified(Recommendation recommendation, OptionSettingItem optionSetting)
+        {
+            // If the option setting is not displayable, that means its dependencies are not satisfied and it does not play any role in the deployment.
+            // We do not need to evaluate whether it has been modified or not.
+            if (!IsOptionSettingDisplayable(recommendation, optionSetting))
+            {
+                return false;
+            }
+
+            if (optionSetting.Type.Equals(OptionSettingValueType.List))
+            {
+                var currentSet = GetOptionSettingValue<SortedSet<string>>(recommendation, optionSetting) ?? new SortedSet<string>();
+                var defaultSet = GetOptionSettingDefaultValue<SortedSet<string>>(recommendation, optionSetting) ?? new SortedSet<string>();
+
+                // return true if both have different lengths or all elements in currentSet are not present in defaultSet
+                return defaultSet.Count != currentSet.Count || !currentSet.All(x => defaultSet.Contains(x));
+            }
+
+            if (optionSetting.Type.Equals(OptionSettingValueType.KeyValue))
+            {
+                var currentDict = GetOptionSettingValue<Dictionary<string, string>>(recommendation, optionSetting) ?? new Dictionary<string, string>();
+                var defaultDict = GetOptionSettingDefaultValue<Dictionary<string, string>>(recommendation, optionSetting) ?? new Dictionary<string, string>();
+
+                // return true if both have different lengths or all keyValue pairs are not equal between currentDict and defaultDict
+                return defaultDict.Count != currentDict.Count ||
```

Comment:
Aren't we missing the case that the keys are all the same but a user changed the value of a key?

Reply:
```
currentDict.All(keyPair => defaultDict.ContainsKey(keyPair.Key) && string.Equals(defaultDict[keyPair.Key], currentDict[keyPair.Key]))
```
The `...&& string.Equals(defaultDict[keyPair.Key], currentDict[keyPair.Key]))` checks that the values are also equal for the same key